File: solutions/fsharp/yacht/1/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One 
    | Two 
    | Three
    | Four 
    | Five 
    | Six

let score category dice =
    let dieToInt = function
        | One -> 1
        | Two -> 2
        | Three -> 3
        | Four -> 4
        | Five -> 5
        | Six -> 6

    let sortedDice = dice |> List.map dieToInt |> List.sort
    let groups = sortedDice |> List.groupBy id |> List.sortBy (snd >> List.length)
    let count x = dice |> List.filter ((=) x) |> List.length

    match category with
    | Ones -> 1 * count One
    | Twos -> 2 * count Two
    | Threes -> 3 * count Three
    | Fours -> 4 * count Four
    | Fives -> 5 * count Five
    | Sixes -> 6 * count Six
    | FullHouse ->
        match groups with
        | [_; (_, threes)] when threes.Length = 3 -> List.sum sortedDice
        | _ -> 0
    | FourOfAKind ->
        match groups with
        | [(value, four)] -> 4 * value
        | [_; (value, four)] when four.Length >= 4 -> 4 * value
        | _ -> 0
    | LittleStraight -> if sortedDice = [1..5] then 30 else 0
    | BigStraight ->    if sortedDice = [2..6] then 30 else 0
    | Choice ->         List.sum sortedDice
    | Yacht ->          if groups.Length = 1 then 50 else 0

File: solutions/haskell/yacht/1/src/Yacht.hs
module Yacht (yacht, Category (..)) where

import qualified Data.List as L

data Category
  = Ones
  | Twos
  | Threes
  | Fours
  | Fives
  | Sixes
  | FullHouse
  | FourOfAKind
  | LittleStraight
  | BigStraight
  | Choice
  | Yacht

yacht :: Category -> [Int] -> Int
yacht category d =
  let dice = L.sort d
      groups = L.sortOn length . L.group $ dice
      count x = length . filter (== x)
   in case category of
        Ones -> 1 * count 1 dice
        Twos -> 2 * count 2 dice
        Threes -> 3 * count 3 dice
        Fours -> 4 * count 4 dice
        Fives -> 5 * count 5 dice
        Sixes -> 6 * count 6 dice
        FullHouse -> case length groups of
          2 -> if length (last groups) == 3 then sum dice else 0
          _ -> 0
        FourOfAKind -> case length groups of
          1 -> 4 * head dice
          2 -> if length (last groups) == 4 then sum . last $ groups else 0
          _ -> 0
        LittleStraight -> if dice == [1 .. 5] then 30 else 0
        BigStraight -> if dice == [2 .. 6] then 30 else 0
        Choice -> sum dice
        Yacht -> if length groups == 1 then 50 else 0
